state/logs.go

```diff
@@ -216,20 +216,50 @@ func NewDbLogger(st ModelSessioner) *DbLogger {
 	}
 }
 
-// Log writes a log message to the database.
-func (logger *DbLogger) Log(t time.Time, entity string, module string, location string, level loggo.Level, msg string) error {
-	// TODO(ericsnow) Use a controller-global int sequence for Id.
-
-	unixEpochNanoUTC := t.UnixNano()
-	return logger.logsColl.Insert(&logDoc{
-		Id:       bson.NewObjectId(),
-		Time:     unixEpochNanoUTC,
-		Entity:   entity,
-		Module:   module,
-		Location: location,
-		Level:    int(level),
-		Message:  msg,
-	})
+// Log writes log messages to the database. Log records
+// are written to the database in bulk; callers should
+// buffer log records to and call Log with a batch to
+// minimise database writes.
+//
+// The ModelUUID and ID fields of records are ignored;
+// DbLogger is scoped to a single model, and ID is
+// controlled by the DbLogger code.
+func (logger *DbLogger) Log(records []LogRecord) error {
+	for _, r := range records {
+		if err := validateInputLogRecord(r); err != nil {
+			return errors.Annotate(err, "validating input log record")
+		}
+	}
+	// TODO(axw) copy session here and close after?
+	bulk := logger.logsColl.Bulk()
```

Comment:
If we have a separate logger per State object, I think re-using just the same session is better than scaling sessions per Log calls.